File: src/melopa/downsample_test.py
"""Tests for plotting downsample algorithms."""

import numpy
from numpy import testing

from melopa import downsample


def test_decimate_evenly_selects_array_elements() -> None:
    """Decimation selects every nth element with index rounding."""
    y = numpy.arange(16)
    actual = downsample.decimate(y, 5)
    testing.assert_array_equal(actual, [0, 4, 8, 11, 15])


def test_lttb_select_first_index() -> None:
    """LTTB selects first index for each bin on a straight line."""
    y = numpy.arange(10)
    actual = downsample.lttb(y, 4)
    testing.assert_array_equal(actual, [0, 1, 5, 9])


def test_lttb_select_single_peak() -> None:
    """LTTB selects single peak."""
    x = numpy.arange(5)
    y = numpy.array([0, 0, 1, 0, 0])
    array = numpy.array([x, y], dtype=y.dtype).T
    actual = downsample.lttb(array, 3)
    testing.assert_array_equal(actual, [0, 2, 4])

File: src/melopa/downsample.test.ts
/**
 * Tests for library functions.
 */

import { expect, test } from "vitest";
import { decimate, lttb } from "./downsample";

test("Decimate evenly selects array elements", () => {
  const x = new Float32Array([...Array(16).keys()]);
  const actual = decimate(x, x, 5)[1];
  expect(actual).toStrictEqual(new Float32Array([0, 4, 8, 11, 15]));
});

test("LTTB selects first index for each bin on a straight line", () => {
  const y = new Float32Array([...Array(10)].keys());
  const actual = lttb(y, y, 4)[1];
  expect(actual).toStrictEqual(new Float32Array([0, 1, 5, 9]));
});

test("LTTB selects single peak", () => {
  const x = new Float32Array([...Array(5)].keys());
  const y = new Float32Array([0, 0, 1, 0, 0]);
  const actual = lttb(x, y, 3)[0];
  expect(actual).toStrictEqual(new Float32Array([0, 2, 4]));
});
